lib/stdlib/test/property_test/queue_prop.erl

```diff
@@ -0,0 +1,684 @@
+                    equal(tl(L), Q1);
+                {empty, Q1} ->
+                    L =:= [] andalso
+                    equal(L, Q1)
+            end
+        end
+    ).
+
+prop_out_r() ->
+    ?FORALL(
+        {L, Q},
+        list_queue(),
+        begin
+            case queue:out_r(Q) of
+                {{value, I}, Q1} ->
+                    L1 = lists:reverse(L),
+                    I =:= hd(L1) andalso
+                    equal(lists:reverse(tl(L1)), Q1);
+                {empty, Q1} ->
+                    L =:= [] andalso
+                    equal(L, Q1)
+            end
+        end
+    ).
+
+prop_peek() ->
+    ?FORALL(
+        {L, Q},
+        list_queue(),
+        begin
+            case queue:peek(Q) of
+                {value, I} ->
+                    I =:= hd(L);
+                empty ->
+                    L =:= []
+            end
+        end
+    ).
+
+prop_peek_r() ->
+    ?FORALL(
+        {L, Q},
+        list_queue(),
+        begin
+            case queue:peek_r(Q) of
+                {value, I} ->
+                    I =:= lists:last(L);
+                empty ->
+                    L =:= []
+            end
+        end
+    ).
+
+prop_reverse() ->
+    ?FORALL(
+        {L, Q},
+        list_queue(),
+        begin
+            equal(lists:reverse(L), queue:reverse(Q))
+        end
+    ).
+
+prop_snoc() ->
+    ?FORALL(
+        L,
+        list(),
+        begin
+            Q = lists:foldl(
+                fun (I, Acc) ->
+                    queue:snoc(Acc, I)
+                end,
+                queue:new(),
+                L
+            ),
+            equal(L, Q)
+        end
+    ).
+
+prop_split() ->
+    ?FORALL(
+        {N, {L, Q}},
+        {non_neg_integer(), list_queue()},
+        begin
+            N1 = N rem (length(L) + 1),
+            {Q1, Q2} = queue:split(N1, Q),
+            {L1, L2} = lists:split(N1, L),
+
+            equal(L1, Q1) andalso
+            equal(L2, Q2)
+        end
+    ).
+
+prop_tail() ->
+    common_drop_tail(fun queue:tail/1).
+
+% Test sequences of insert and retrieval operations
+prop_ops() ->
+    ?FORALL(
+        {Ops, {L, Q}},
+        {
+            list(
+                oneof([{cons, term()},
+                       daeh,
+                       drop,
+                       drop_r,
+                       get,
+                       get_r,
+                       head,
+                       {in, term()},
+                       {in_r, term()},
+                       init,
+                       liat,
+                       last,
+                       out,
+                       out_r,
+                       peek,
+                       peek_r,
+                       {snoc, term()},
+                       tail])
+            ),
+            list_queue()
+        },
+        begin
+            {Res, RQ, RL} = lists:foldl(
+                fun
+                    (_, Acc = {false, _, _}) ->
```

Comment:
Good with a test to test a sequence of ops and not just individual operations.

For the sake of readability, I see no reason to have this as an inlined fun.  There are no free names in the fun, so it can be written as a named function instead.  This also means less indentation.
This test is about randomly creating a sequence of operations, applying them to  the `queue` implementation and to a reference implementation (model) and the comparing the results.  For the sake of decreasing complexity one could split these into different parts:
1. The simple reference implementation/model using lists.  This can be done in one function taking two args: operation as generated and the _queue as list_, returning values similar to the implementation to be tested.
2. A mapping from generated operations to `queue` operations, so applying a generated operation is simple
3. A general function for comparing the results of of operations, i.e., that returned values are the "same" and that the model and implementation correspond to each other.
I _think_ that this would be more readable than the current implementation since it does not require looking at every clause and figure out what the test is.  I will admit this is a bit thinking out loud right now and understand if it is vague.  If needed, I can take some time to flesh out some code to clarify what I mean.

Reply:
I'm currently drawing up something that does what I think you want. But it seems to turn out in a way that doesn't really reduce the overall complexity IMO. It's not finished yet, probably tomorrow.

Reply:
See latest commit for my suggestion. As I said, it is IMO more complex and way more verbose than the previous implementation.

Reply:
I really don't agree that it is more complex.  It has the advantage of clearly separating the different parts, i.e., driver, model and implementation.  As for verbosity, possibly yes, but that isn't necessarily bad, especially when it leads to better clarity when reading.
In `2` above I suggested a mapping from generated operations to `queue` operations.  Instead of doing this with a number of clauses and `when` expressions, one can do it directly with a map, i.e.,
```
#{peek => fun queue:peek/1,
   get => fun queue:get/1,
   in => fun queue:in/2,
  ... etc
}
```
Preserve the return values and exceptions and have the list model have the same semantics and return values instead of writing scaffolding around the functions in `queue` or write a general conversation for the case when `error:empty` is raised.  You'll need some additional code to take care of the difference when operations have 1 or two arguments, but that should be quite simple.  The main point is making the mapping from generated operations to `queue` operations so simple that one does not question them and then have a simple list model for confirmation.

I do think that what you have now is better than your original implementation, but I also think that it can be made even better.

Reply:
I'll have to think about this a bit and will come back to it next week  :sweat_smile:

Reply:
Hm, I have been thinking about this time and again, but I can't get to a satisfactory result. I guess I'm stuck (or dense :sweat:)
To be clear, I take it that you think of running the list of generated operations with the queue and with the model, then compare the collected return values and final states of each, right?

The mapping approach for (2) is problematic since not all queue functions have the same argument order. Specifically, `snoc` is an exception because it takes the queue as first argument, while all other have it as last. That can be solved by handling that single exception specially, no biggie.

What is more problematic is that without knowing the operation (function) we're running, we don't know how to interpret the return value and what to pass into the next function.
For some functions (`get`, `get_r`, `head` etc), the return value is an item from the queue (or an error exception) and the queue itself is unchanged and what is to be passed on is the queue we already have.
For other functions (`drop`, `drop_r`, `tail` etc), the return value is the changed queue (to be passed on) itself, _unless_ there is an error exception, _then_ the queue we already have needs to be passed on.
Still others (`out` and `out_r`) return _both_ an item _and_ the updated queue, in which case we have to take it apart in order to separate the item from the queue to pass on.

In a nutshell, no matter how go about it, I have to heavily rely on the operation, which leads me back to just another implementation of what I have already. I could strip away some of the wrappings I did, yes, but it always boils back down to the same or similar set of (function or case) clauses.

Reply:
It's just me being lazy, describing in (too) abstract terms instead of writing code that shows what I was thinking about.  Below are two functions; one for constructing the mapping and a rewritten `do_op_queue/2` that uses the mapping.
I have preserved the call semantics that you had, so it is only the expression of the mapping that is different.  By setting up some higher order functions to facilitate the construction of the map, the actual mapping can then be expressed rather compact, hopefully making it easier to understand.
As stated in the comments, it could be made simpler, but that would affect the semantics, so further changes elsewhere, i.e., in `do_op_list/2` would be needed.  To be honest, I would actually rewrite that function as well not only for the reason of making the semantics simpler, but for clarity.  I think it would be easier if grouped by operation rather than argument.  Currently, several operations are described by two clauses far from each other, so to understand, e.g., `init`, one has to find the two clauses where `init` is in the guard instead of finding one clause (or mapping) where all of `init` is described.  Yes, this is me being lazy again and describing in abstract terms :smile: .. sorry. 

```
%% Construct the mapping of an abstract operation Op | {Op, I} to a
%% function that always returns
%%  {{value, <value>} |  undefined | empty, <queue>}
%% Written this way to not rewrite do_op_list/2 (which also could be
%% simplified).
queue_map() ->
    %% Swap arguments
    Swap = fun(F) -> fun(I, Q) -> F(Q, I) end end,

    %% Affects only queue, returns no value, one or two args
    QOp1 = fun(F) -> fun(Q) -> {undefined, F(Q)} end end,
    QOp2 = fun(F) -> fun(I, Q) -> {undefined, F(I, Q)} end end,
    %% Returns value, does not affect queue.
    %% For functions that return {value, <value>} | empty
    IValue = fun(F) -> fun(Q) -> {F(Q), Q} end end,
    %% For functions thath just return the value
    Value = fun(F) -> fun(Q) -> {{value, F(Q)}, Q} end end,

    %% Catch error and map to {ErrTo, <unchanged queue>}
    %% If we mapped to {empty, ..} in both cases we could remove the
    %% second argument and have no need for both VCatch and UCatch
    Catch = fun(F, ErrTo) -> fun(Q) ->
                                     try F(Q)
                                     catch
                                         error:empty ->
                                             {ErrTo, Q}
                                     end
                             end
            end,

    VCatch = fun(F) -> Catch(F, empty) end,
    UCatch = fun(F) -> Catch(F, undefined) end,

    #{in => QOp2(fun queue:in/2),
      in_r => QOp2(fun queue:in_r/2),
      cons => QOp2(fun queue:cons/2),
      snoc => QOp2(Swap(fun queue:snoc/2)),
      out => fun queue:out/1,
      out_r => fun queue:out_r/1,
      peek => IValue(fun queue:peek/1),
      peek_r => IValue(fun queue:peek_r/1),

      get => VCatch(Value(fun queue:get/1)),
      get_r => VCatch(Value(fun queue:get_r/1)),
      head => VCatch(Value(fun queue:head/1)),
      daeh => VCatch(Value(fun queue:daeh/1)),
      last => VCatch(Value(fun queue:last/1)),

      init => UCatch(QOp1(fun queue:init/1)),
      drop => UCatch(QOp1(fun queue:drop/1)),
      drop_r => UCatch(QOp1(fun queue:drop_r/1)),
      tail => UCatch(QOp1(fun queue:tail/1)),
      liat => UCatch(QOp1(fun queue:liat/1))
     }.

do_op_queue({Op, I}, Q) ->
    F = maps:get(Op, queue_map()),
    F(I, Q);
do_op_queue(Op, Q) ->
    F = maps:get(Op, queue_map()),
    F(Q).
```

Reply:
:smiley_cat: Ok, I'll try to work with that, thanks :wink:

Reply:
Ok, please see my last commit, is that better? :blush: I took most of your example and made some minor changes. I'll add an explanatory comment if you say that it's to your liking :wink:

Reply:
Yes, this is much better!